File: frontend/src/types/hud.ts
/**
 * HUD Compliance TypeScript Interfaces
 * Type definitions for HUD PRAC compliance models matching backend structure
 */

// Enums
export type CertificationType = 'initial' | 'annual' | 'interim' | 'other';
export type CertificationStatus = 'pending' | 'approved' | 'rejected' | 'submitted';
export type RelationshipType = 'head' | 'spouse' | 'child' | 'other_dependent' | 'live_in_aide' | 'other';
export type IncomeType = 'wages' | 'self_employment' | 'social_security' | 'ssi' | 'pension' | 'child_support' | 'alimony' | 'unemployment' | 'tanf' | 'other';
export type VerificationType = 'pay_stub' | 'tax_return' | 'award_letter' | 'employer_statement' | 'bank_statement' | 'self_certification';
export type InspectionType = 'initial' | 'annual' | 'complaint' | 'special';
export type InspectionStatus = 'passed' | 'failed' | 'conditional' | 'pending';

// Base interfaces
export interface BaseEntity {
  id: string;
  created_at: string;
  updated_at?: string;
}

export interface BaseOrgEntity extends BaseEntity {
  org_id: string;
}

// Tenant Income Certification interface
export interface TenantIncomeCertification extends BaseOrgEntity {
  tenant_id: string;
  property_id: string;
  unit_id?: string;
  certification_date: string;
  effective_date: string;
  cert_type: CertificationType;
  household_size: number;
  annual_income: string; // Decimal as string
  adjusted_income: string; // Decimal as string
  tenant_rent_portion: string; // Decimal as string
  utility_allowance: string; // Decimal as string
  subsidy_amount: string; // Decimal as string
  certification_status: CertificationStatus;
  hud_50059_submitted: boolean;
  hud_50059_submission_date?: string;
  created_by: string;
}

// Household Member interface
export interface HouseholdMember extends BaseEntity {
  tic_id: string;
  full_name: string;
  ssn_last_4?: string;
  date_of_birth: string;
  relationship_type: RelationshipType;
  is_student: boolean;
  is_disabled: boolean;
  annual_income: string; // Decimal as string
}

// Income Source interface
export interface IncomeSource extends BaseEntity {
  household_member_id: string;
  income_type: IncomeType;
  employer_name?: string;
  monthly_amount: string; // Decimal as string
  annual_amount: string; // Decimal as string
  verification_type: VerificationType;
  verification_date: string;
}

// Utility Allowance interface
export interface UtilityAllowance extends BaseOrgEntity {
  property_id: string;
  bedroom_count: number;
  heating: string; // Decimal as string
  cooking: string; // Decimal as string
  lighting: string; // Decimal as string
  water_sewer: string; // Decimal as string
  trash: string; // Decimal as string
  total_allowance: string; // Decimal as string
  effective_date: string;
}

// REAC Inspection interface
export interface REACInspection extends BaseEntity {
  property_id: string;
  inspection_date: string;
  inspection_type: InspectionType;
  overall_score?: number;
  inspection_status: InspectionStatus;
  deficiencies_count: number;
  critical_deficiencies: number;
  report_url?: string;
  next_inspection_date?: string;
}

// Extended interfaces with relationships
export interface TenantIncomeCertificationWithMembers extends TenantIncomeCertification {
  household_members: HouseholdMember[];
}

export interface HouseholdMemberWithIncomeSources extends HouseholdMember {
  income_sources: IncomeSource[];
}

export interface TenantIncomeCertificationComplete extends TenantIncomeCertification {
  household_members: HouseholdMemberWithIncomeSources[];
}

// API Response interfaces
export interface ApiResponse<T> {
  data: T;
}

export interface ApiListResponse<T> {
  data: T[];
}

// Request interfaces for creating/updating entities
export interface CreateTenantIncomeCertificationRequest {
  tenant_id: string;
  property_id: string;
  unit_id?: string;
  certification_date: string;
  effective_date: string;
  cert_type: CertificationType;
  household_size: number;
  annual_income: string;
  adjusted_income: string;
  tenant_rent_portion: string;
  utility_allowance: string;
  subsidy_amount: string;
  certification_status?: CertificationStatus;
}

export interface UpdateTenantIncomeCertificationRequest {
  tenant_id?: string;
  property_id?: string;
  unit_id?: string;
  certification_date?: string;
  effective_date?: string;
  cert_type?: CertificationType;
  household_size?: number;
  annual_income?: string;
  adjusted_income?: string;
  tenant_rent_portion?: string;
  utility_allowance?: string;
  subsidy_amount?: string;
  certification_status?: CertificationStatus;
}

export interface CreateHouseholdMemberRequest {
  full_name: string;
  ssn_last_4?: string;
  date_of_birth: string;
  relationship_type: RelationshipType;
  is_student?: boolean;
  is_disabled?: boolean;
  annual_income?: string;
}

export interface UpdateHouseholdMemberRequest {
  full_name?: string;
  ssn_last_4?: string;
  date_of_birth?: string;
  relationship_type?: RelationshipType;
  is_student?: boolean;
  is_disabled?: boolean;
  annual_income?: string;
}

export interface CreateIncomeSourceRequest {
  income_type: IncomeType;
  employer_name?: string;
  monthly_amount: string;
  annual_amount: string;
  verification_type: VerificationType;
  verification_date: string;
}

export interface UpdateIncomeSourceRequest {
  income_type?: IncomeType;
  employer_name?: string;
  monthly_amount?: string;
  annual_amount?: string;
  verification_type?: VerificationType;
  verification_date?: string;
}

export interface CreateUtilityAllowanceRequest {
  property_id: string;
  bedroom_count: number;
  heating?: string;
  cooking?: string;
  lighting?: string;
  water_sewer?: string;
  trash?: string;
  total_allowance: string;
  effective_date: string;
}

export interface UpdateUtilityAllowanceRequest {
  property_id?: string;
  bedroom_count?: number;
  heating?: string;
  cooking?: string;
  lighting?: string;
  water_sewer?: string;
  trash?: string;
  total_allowance?: string;
  effective_date?: string;
}

export interface CreateREACInspectionRequest {
  property_id: string;
  inspection_date: string;
  inspection_type: InspectionType;
  overall_score?: number;
  inspection_status: InspectionStatus;
  deficiencies_count?: number;
  critical_deficiencies?: number;
  report_url?: string;
  next_inspection_date?: string;
}

export interface UpdateREACInspectionRequest {
  property_id?: string;
  inspection_date?: string;
  inspection_type?: InspectionType;
  overall_score?: number;
  inspection_status?: InspectionStatus;
  deficiencies_count?: number;
  critical_deficiencies?: number;
  report_url?: string;
  next_inspection_date?: string;
}

// Filter interfaces for API queries
export interface CertificationFilters {
  property_id?: string;
  tenant_id?: string;
  status?: CertificationStatus;
  cert_type?: CertificationType;
}

export interface HouseholdMemberFilters {
  tic_id?: string;
  relationship_type?: RelationshipType;
}

export interface IncomeSourceFilters {
  household_member_id?: string;
  income_type?: IncomeType;
  verification_type?: VerificationType;
}

export interface UtilityAllowanceFilters {
  property_id?: string;
  bedroom_count?: number;
}

export interface REACInspectionFilters {
  property_id?: string;
  inspection_type?: InspectionType;
  inspection_status?: InspectionStatus;
}

// Special request interfaces
export interface SubmitHUD50059Request {
  // No additional fields needed - just the cert_id in URL
}

export interface CalculateRentRequest {
  household_income: string;
  household_size: number;
  utility_allowance: string;
  contract_rent: string;
}

export interface CalculateRentResponse {
  tenant_rent: string;
  subsidy_amount: string;
  total_contract_rent: string;
  utility_allowance: string;
  monthly_income: string;
  rent_to_income_ratio: string;
}

// Expiring/Upcoming query interfaces
export interface ExpiringCertificationsRequest {
  days?: number; // Default 30
}

export interface UpcomingInspectionsRequest {
  days?: number; // Default 60
}

// Current allowance request
export interface CurrentUtilityAllowanceRequest {
  property_id: string;
  bedroom_count: number;
}

// Dashboard/Summary interfaces
export interface CertificationSummary {
  total_certifications: number;
  pending_certifications: number;
  approved_certifications: number;
  expiring_soon: number;
  overdue_certifications: number;
}

export interface InspectionSummary {
  total_inspections: number;
  passed_inspections: number;
  failed_inspections: number;
  conditional_inspections: number;
  upcoming_inspections: number;
  average_score: number;
}

export interface ComplianceDashboard {
  certification_summary: CertificationSummary;
  inspection_summary: InspectionSummary;
  recent_certifications: TenantIncomeCertification[];
  upcoming_inspections: REACInspection[];
  expiring_certifications: TenantIncomeCertification[];
}

// Report interfaces
export interface CertificationReport {
  certification_id: string;
  tenant_name: string;
  property_name: string;
  unit_number?: string;
  certification_date: string;
  effective_date: string;
  cert_type: CertificationType;
  household_size: number;
  annual_income: string;
  adjusted_income: string;
  tenant_rent_portion: string;
  utility_allowance: string;
  subsidy_amount: string;
  certification_status: CertificationStatus;
  hud_50059_submitted: boolean;
  hud_50059_submission_date?: string;
  days_until_expiry: number;
}

export interface InspectionReport {
  inspection_id: string;
  property_name: string;
  inspection_date: string;
  inspection_type: InspectionType;
  overall_score?: number;
  inspection_status: InspectionStatus;
  deficiencies_count: number;
  critical_deficiencies: number;
  report_url?: string;
  next_inspection_date?: string;
  days_until_next: number;
}

export interface UtilityAllowanceReport {
  property_name: string;
  bedroom_count: number;
  heating: string;
  cooking: string;
  lighting: string;
  water_sewer: string;
  trash: string;
  total_allowance: string;
  effective_date: string;
  units_affected: number;
}

// Form validation interfaces
export interface CertificationFormData {
  tenant_id: string;
  property_id: string;
  unit_id?: string;
  certification_date: string;
  effective_date: string;
  cert_type: CertificationType;
  household_size: number;
  annual_income: string;
  adjusted_income: string;
  tenant_rent_portion: string;
  utility_allowance: string;
  subsidy_amount: string;
  household_members: HouseholdMemberFormData[];
}

export interface HouseholdMemberFormData {
  full_name: string;
  ssn_last_4?: string;
  date_of_birth: string;
  relationship_type: RelationshipType;
  is_student: boolean;
  is_disabled: boolean;
  annual_income: string;
  income_sources: IncomeSourceFormData[];
}

export interface IncomeSourceFormData {
  income_type: IncomeType;
  employer_name?: string;
  monthly_amount: string;
  annual_amount: string;
  verification_type: VerificationType;
  verification_date: string;
}

export interface UtilityAllowanceFormData {
  property_id: string;
  bedroom_count: number;
  heating: string;
  cooking: string;
  lighting: string;
  water_sewer: string;
  trash: string;
  total_allowance: string;
  effective_date: string;
}

export interface REACInspectionFormData {
  property_id: string;
  inspection_date: string;
  inspection_type: InspectionType;
  overall_score?: number;
  inspection_status: InspectionStatus;
  deficiencies_count: number;
  critical_deficiencies: number;
  report_url?: string;
  next_inspection_date?: string;
}

File: backend/app/models/hud.py
"""
HUD PRAC Compliance Models
SQLAlchemy models for HUD compliance tracking and reporting
"""
from sqlalchemy import Column, String, Integer, Boolean, DateTime, Date, Numeric, ForeignKey, Index, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from sqlalchemy.ext.declarative import declarative_base
from datetime import datetime
from uuid import uuid4
from enum import Enum as PyEnum

Base = declarative_base()


class CertificationType(str, PyEnum):
    """Types of income certifications"""
    INITIAL = "initial"
    ANNUAL = "annual"
    INTERIM = "interim"
    OTHER = "other"


class CertificationStatus(str, PyEnum):
    """Status of income certification"""
    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"


class RelationshipType(str, PyEnum):
    """Household member relationship types"""
    HEAD = "head"
    SPOUSE = "spouse"
    CHILD = "child"
    OTHER = "other"


class IncomeType(str, PyEnum):
    """Types of income sources"""
    WAGES = "wages"
    SALARY = "salary"
    SOCIAL_SECURITY = "social_security"
    SSI = "ssi"
    SSDI = "ssdi"
    UNEMPLOYMENT = "unemployment"
    WORKERS_COMP = "workers_comp"
    CHILD_SUPPORT = "child_support"
    ALIMONY = "alimony"
    PENSION = "pension"
    ANNUITY = "annuity"
    INTEREST = "interest"
    DIVIDENDS = "dividends"
    CAPITAL_GAINS = "capital_gains"
    BUSINESS_INCOME = "business_income"
    RENTAL_INCOME = "rental_income"
    OTHER = "other"


class VerificationType(str, PyEnum):
    """Types of income verification"""
    PAY_STUB = "pay_stub"
    TAX_RETURN = "tax_return"
    AWARD_LETTER = "award_letter"
    BANK_STATEMENT = "bank_statement"
    EMPLOYER_VERIFICATION = "employer_verification"
    AGENCY_VERIFICATION = "agency_verification"
    OTHER = "other"


class InspectionType(str, PyEnum):
    """Types of REAC inspections"""
    INITIAL = "initial"
    ANNUAL = "annual"
    COMPLAINT = "complaint"
    FOLLOW_UP = "follow_up"


class InspectionStatus(str, PyEnum):
    """Status of REAC inspection"""
    PASSED = "passed"
    FAILED = "failed"
    CONDITIONAL = "conditional"
    PENDING = "pending"


class TenantIncomeCertification(Base):
    """HUD Tenant Income Certification (TIC) records"""
    __tablename__ = "tenant_income_certifications"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    org_id = Column(UUID(as_uuid=True), ForeignKey("organizations.id"), nullable=False, index=True)
    tenant_id = Column(UUID(as_uuid=True), ForeignKey("tenants.id"), nullable=False, index=True)
    property_id = Column(UUID(as_uuid=True), ForeignKey("properties.id"), nullable=False, index=True)
    unit_id = Column(UUID(as_uuid=True), ForeignKey("units.id"), nullable=True, index=True)
    
    # Certification details
    certification_date = Column(Date, nullable=False)
    effective_date = Column(Date, nullable=False)
    cert_type = Column(String(20), nullable=False)  # CertificationType enum
    household_size = Column(Integer, nullable=False)
    
    # Income calculations
    annual_income = Column(Numeric(12, 2), nullable=False)
    adjusted_income = Column(Numeric(12, 2), nullable=False)
    tenant_rent_portion = Column(Numeric(12, 2), nullable=False)
    utility_allowance = Column(Numeric(12, 2), nullable=False)
    subsidy_amount = Column(Numeric(12, 2), nullable=False)
    
    # Status and compliance
    certification_status = Column(String(20), nullable=False, default="pending")  # CertificationStatus enum
    hud_50059_submitted = Column(Boolean, default=False)
    hud_50059_submission_date = Column(DateTime, nullable=True)
    
    # Audit fields
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, onupdate=datetime.utcnow)
    deleted_at = Column(DateTime, nullable=True)
    
    # Relationships
    organization = relationship("Organization")
    tenant = relationship("Tenant")
    property = relationship("Property")
    unit = relationship("Unit")
    creator = relationship("User")
    household_members = relationship("HouseholdMember", back_populates="certification", cascade="all, delete-orphan")
    
    # Indexes
    __table_args__ = (
        Index("idx_tic_org_tenant", "org_id", "tenant_id"),
        Index("idx_tic_org_property", "org_id", "property_id"),
        Index("idx_tic_effective_date", "effective_date"),
        Index("idx_tic_status", "certification_status"),
        Index("idx_tic_type", "cert_type"),
    )


class HouseholdMember(Base):
    """Household members for income certification"""
    __tablename__ = "household_members"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    tic_id = Column(UUID(as_uuid=True), ForeignKey("tenant_income_certifications.id"), nullable=False, index=True)
    
    # Personal information
    full_name = Column(String(255), nullable=False)
    ssn_last_4 = Column(String(4), nullable=True)
    date_of_birth = Column(Date, nullable=False)
    relationship_type = Column(String(20), nullable=False)  # RelationshipType enum
    
    # Status flags
    is_student = Column(Boolean, default=False)
    is_disabled = Column(Boolean, default=False)
    
    # Income
    annual_income = Column(Numeric(12, 2), nullable=False, default=0)
    
    # Timestamps
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, onupdate=datetime.utcnow)
    deleted_at = Column(DateTime, nullable=True)
    
    # Relationships
    certification = relationship("TenantIncomeCertification", back_populates="household_members")
    income_sources = relationship("IncomeSource", back_populates="household_member", cascade="all, delete-orphan")
    
    # Indexes
    __table_args__ = (
        Index("idx_hm_tic", "tic_id"),
        Index("idx_hm_relationship_type", "relationship_type"),
    )


class IncomeSource(Base):
    """Income sources for household members"""
    __tablename__ = "income_sources"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    household_member_id = Column(UUID(as_uuid=True), ForeignKey("household_members.id"), nullable=False, index=True)
    
    # Income details
    income_type = Column(String(30), nullable=False)  # IncomeType enum
    employer_name = Column(String(255), nullable=True)
    monthly_amount = Column(Numeric(12, 2), nullable=False)
    annual_amount = Column(Numeric(12, 2), nullable=False)
    
    # Verification
    verification_type = Column(String(30), nullable=False)  # VerificationType enum
    verification_date = Column(Date, nullable=False)
    
    # Timestamps
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, onupdate=datetime.utcnow)
    deleted_at = Column(DateTime, nullable=True)
    
    # Relationships
    household_member = relationship("HouseholdMember", back_populates="income_sources")
    
    # Indexes
    __table_args__ = (
        Index("idx_is_hm", "household_member_id"),
        Index("idx_is_type", "income_type"),
        Index("idx_is_verification", "verification_type"),
    )


class UtilityAllowance(Base):
    """Utility allowances by bedroom count and property"""
    __tablename__ = "utility_allowances"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    org_id = Column(UUID(as_uuid=True), ForeignKey("organizations.id"), nullable=False, index=True)
    property_id = Column(UUID(as_uuid=True), ForeignKey("properties.id"), nullable=False, index=True)
    
    # Allowance details
    bedroom_count = Column(Integer, nullable=False)
    heating = Column(Numeric(8, 2), nullable=False, default=0)
    cooking = Column(Numeric(8, 2), nullable=False, default=0)
    lighting = Column(Numeric(8, 2), nullable=False, default=0)
    water_sewer = Column(Numeric(8, 2), nullable=False, default=0)
    trash = Column(Numeric(8, 2), nullable=False, default=0)
    total_allowance = Column(Numeric(8, 2), nullable=False)
    
    # Effective date
    effective_date = Column(Date, nullable=False)
    
    # Timestamps
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, onupdate=datetime.utcnow)
    deleted_at = Column(DateTime, nullable=True)
    
    # Relationships
    organization = relationship("Organization")
    property = relationship("Property")
    
    # Indexes
    __table_args__ = (
        Index("idx_ua_org_property", "org_id", "property_id"),
        Index("idx_ua_bedrooms", "bedroom_count"),
        Index("idx_ua_effective_date", "effective_date"),
    )


class REACInspection(Base):
    """REAC (Real Estate Assessment Center) inspection records"""
    __tablename__ = "reac_inspections"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    property_id = Column(UUID(as_uuid=True), ForeignKey("properties.id"), nullable=False, index=True)
    
    # Inspection details
    inspection_date = Column(Date, nullable=False)
    inspection_type = Column(String(20), nullable=False)  # InspectionType enum
    overall_score = Column(Integer, nullable=True)  # 0-100 scale
    inspection_status = Column(String(20), nullable=False)  # InspectionStatus enum
    
    # Deficiency tracking
    deficiencies_count = Column(Integer, nullable=False, default=0)
    critical_deficiencies = Column(Integer, nullable=False, default=0)
    
    # Documentation
    report_url = Column(String(500), nullable=True)
    next_inspection_date = Column(Date, nullable=True)
    
    # Timestamps
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, onupdate=datetime.utcnow)
    deleted_at = Column(DateTime, nullable=True)
    
    # Relationships
    property = relationship("Property")
    
    # Indexes
    __table_args__ = (
        Index("idx_reac_property", "property_id"),
        Index("idx_reac_date", "inspection_date"),
        Index("idx_reac_type", "inspection_type"),
        Index("idx_reac_status", "inspection_status"),
        Index("idx_reac_score", "overall_score"),
    )
